file/sequence_file_reader.cc

```diff
@@ -25,19 +26,32 @@ namespace ROCKSDB_NAMESPACE {
 IOStatus SequentialFileReader::Create(
     const std::shared_ptr<FileSystem>& fs, const std::string& fname,
     const FileOptions& file_opts, std::unique_ptr<SequentialFileReader>* reader,
-    IODebugContext* dbg) {
+    IODebugContext* dbg, RateLimiter* rate_limiter) {
   std::unique_ptr<FSSequentialFile> file;
   IOStatus io_s = fs->NewSequentialFile(fname, file_opts, &file, dbg);
   if (io_s.ok()) {
-    reader->reset(new SequentialFileReader(std::move(file), fname));
+    reader->reset(new SequentialFileReader(std::move(file), fname, nullptr, {},
+                                           rate_limiter));
   }
   return io_s;
 }
 
-IOStatus SequentialFileReader::Read(size_t n, Slice* result, char* scratch) {
+IOStatus SequentialFileReader::Read(size_t n, Slice* result, char* scratch,
+                                    Env::IOPriority rate_limiter_priority) {
   IOStatus io_s;
+  uint64_t filesize;
+  IOOptions io_opts;
+  // Need to get file size for each read since new content might be added to
+  // file after last read
+  io_s = file_->GetFileSize(filesize);
+  if (!io_s.ok()) {
+    return io_s;
+  }
+  // Cap the bytes limit to number of remaining bytes in the file
+  n = std::min(n, static_cast<size_t>(filesize) - offset_);
```

Comment:
Should it be ...offset_.load() since is a atomic var? I am surprised that type check does not catch anything for your current usage (i.e, "-" operation between a size_t and a atomic<size_t> 🤔

Another thing I'd do is to `assert(static_cast<size_t>(filesize) >= offset_.load())` (which only takes effects in non-prod env) as comparing a (theoretically possible but not practically possible as `offset_` should not be greater than the `filesize`, assuming both variables stay correct in impl) negative number with `size_t n` (non-negative) can give surprising casting behavior.

Reply:
Good catch! Is SequentialFileReader expected to be thread-safe btw?

Reply:
> Should it be ...offset_.load() since it's an atomic var? I am surprised that type check does not catch anything for your current usage (i.e, "-" operation between a size_t and a atomic<size_t> 🤔

This usage does the same thing.

Reply:
I was under impression that atomic_var does not have a `-` operation defined looking at https://en.cppreference.com/w/cpp/atomic/atomic so I assume this will need to go through some casting that isn't standard (atomic<size_t> -> size_t). 

I'd love to know the reference to "This usage does the same thing." as it's not obvious to me :p

Reply:
Maybe these two? https://en.cppreference.com/w/cpp/atomic/atomic/operator%3D
https://en.cppreference.com/w/cpp/atomic/atomic/operator_T

Reply:
yeah probably implicit conversion (https://en.cppreference.com/w/cpp/atomic/atomic/operator_T) to pass to operator+(size_t, size_t). I don't plan to confirm it.

Reply:
@cbi42 @ajkr  thanks - then both ways are good with me!